# CNVPaneOfNormal

![Last commit](https://img.shields.io/github/last-commit/lindenb/jvarkit.png)

Call CNV from panel of normal computed with 'basecoverage'


## Usage

```
Usage: java -jar dist/cnvpanelofnormal.jar  [options] Files
Usage: cnvpanelofnormal [options] Files
  Options:
    --bcf-output
      If this program writes a VCF to a file, The format is first guessed from 
      the file suffix. Otherwise, force BCF output. The current supported BCF 
      version is : 2.1 which is not compatible with bcftools/htslib (last 
      checked 2019-11-15)
      Default: false
    --canvas
      Canvas output
    --generate-vcf-md5
      Generate MD5 checksum for VCF output.
      Default: false
    -h, --help
      print help and exit
    --helpFormat
      What kind of help. One of [usage,markdown,xml].
    -f, --iqr
      IQR factor to detect outliers
      Default: 1.5
    --mapq
       min mapping quality.
      Default: 1
  * -V, --vcf, --normal
      Vcf panel of normal generated by 'basecoverage'
    -o, --out
      Output file. Optional . Default: stdout
    -R, --reference
      Indexed fasta Reference file. This file must be indexed with samtools 
      faidx and with picard/gatk CreateSequenceDictionary or samtools dict
    --runmed
       moving median size
      Default: 31
    -t, --treshold
      CNV treshold DEL if depth < 1.0 - x and DUP  if depth > 1.0 + x
      Default: 0.4
    --version
      print version and exit
    -S
      merge CNV within distance of 'x'.
      Default: 0
    -m
      min CNV size
      Default: 0

```


## Keywords

 * depth
 * bam
 * sam
 * coverage
 * vcf
 * cnv
 * sv


## Compilation

### Requirements / Dependencies

* java [compiler SDK 17](https://jdk.java.net/17/). Please check that this java is in the `${PATH}`. Setting JAVA_HOME is not enough : (e.g: https://github.com/lindenb/jvarkit/issues/23 )


### Download and Compile

```bash
$ git clone --recurse-submodules "https://github.com/lindenb/jvarkit.git"
$ cd jvarkit
$ ./gradlew cnvpanelofnormal
```

The java jar file will be installed in the `dist` directory.


## Creation Date

20241123

## Source code 

[https://github.com/lindenb/jvarkit/tree/master/src/main/java/com/github/lindenb/jvarkit/tools/basecoverage/CNVPaneOfNormal.java](https://github.com/lindenb/jvarkit/tree/master/src/main/java/com/github/lindenb/jvarkit/tools/basecoverage/CNVPaneOfNormal.java)


## Contribute

- Issue Tracker: [http://github.com/lindenb/jvarkit/issues](http://github.com/lindenb/jvarkit/issues)
- Source Code: [http://github.com/lindenb/jvarkit](http://github.com/lindenb/jvarkit)

## License

The project is licensed under the MIT license.

## Citing

Should you cite **cnvpanelofnormal** ? [https://github.com/mr-c/shouldacite/blob/master/should-I-cite-this-software.md](https://github.com/mr-c/shouldacite/blob/master/should-I-cite-this-software.md)

The current reference is:

[http://dx.doi.org/10.6084/m9.figshare.1425030](http://dx.doi.org/10.6084/m9.figshare.1425030)

> Lindenbaum, Pierre (2015): JVarkit: java-based utilities for Bioinformatics. figshare.
> [http://dx.doi.org/10.6084/m9.figshare.1425030](http://dx.doi.org/10.6084/m9.figshare.1425030)


