# SamTranslocations

Explore translocations between two chromosomes using discordant paired-end reads.


## Usage

```
Usage: samtranslocations [options] Files
  Options:
    --filter
      A JEXL Expression that will be used to filter out some sam-records (see 
      https://software.broadinstitute.org/gatk/documentation/article.php?id=1255). 
      An expression should return a boolean value (true=exclude, false=keep 
      the read). An empty expression keeps everything. The variable 'record' 
      is the current observed read, an instance of SAMRecord (https://samtools.github.io/htsjdk/javadoc/htsjdk/htsjdk/samtools/SAMRecord.html).
      Default: record.getMappingQuality()<1 || record.getDuplicateReadFlag() || record.getReadFailsVendorQualityCheckFlag() || record.isSecondaryOrSupplementary()
    --groupby
      Group Reads by. Data partitioning using the SAM Read Group (see 
      https://gatkforums.broadinstitute.org/gatk/discussion/6472/ ) . It can 
      be any combination of sample, library....
      Default: sample
      Possible Values: [readgroup, sample, library, platform, center, sample_by_platform, sample_by_center, sample_by_platform_by_center, any]
    -h, --help
      print help and exit
    --helpFormat
      What kind of help
      Possible Values: [usage, markdown, xml]
    --region, --interval
      Limit analysis to this interval. An interval as the following syntax : 
      "chrom:start-end" or "chrom:middle+extend"  or 
      "chrom:start-end+extend".A program might use a Reference sequence to fix 
      the chromosome name (e.g: 1->chr1)
    --maxRecordsInRam
      When writing  files that need to be sorted, this will specify the number 
      of records stored in RAM before spilling to disk. Increasing this number 
      reduces the number of file  handles needed to sort a file, and increases 
      the amount of RAM needed
      Default: 50000
    -m, --min-events
      Minimal number of events for printing a result
      Default: 1
    -o, --output
      Output file. Optional . Default: stdout
    -r, --round
      Round locations to LOC=LOC-LOC%round
      Default: 150
    --tmpDir
      tmp working directory. Default: java.io.tmpDir
      Default: []
    --version
      print version and exit
    -x, --xml
      XML Output
      Default: false

```


## Keywords

 * sam
 * bam
 * xslt
 * xml


## Compilation

### Requirements / Dependencies

* java [compiler SDK 1.8](http://www.oracle.com/technetwork/java/index.html) (**NOT the old java 1.7 or 1.6**) and avoid OpenJdk, use the java from Oracle. Please check that this java is in the `${PATH}`. Setting JAVA_HOME is not enough : (e.g: https://github.com/lindenb/jvarkit/issues/23 )
* GNU Make >= 3.81
* curl/wget
* git
* xsltproc http://xmlsoft.org/XSLT/xsltproc2.html (tested with "libxml 20706, libxslt 10126 and libexslt 815")


### Download and Compile

```bash
$ git clone "https://github.com/lindenb/jvarkit.git"
$ cd jvarkit
$ make samtranslocations
```

The *.jar libraries are not included in the main jar file, [so you shouldn't move them](https://github.com/lindenb/jvarkit/issues/15#issuecomment-140099011 ).
The required libraries will be downloaded and installed in the `dist` directory.

Experimental: you can also create a [fat jar](https://stackoverflow.com/questions/19150811/) which contains classes from all the libraries, on which your project depends (it's bigger). Those fat-jar are generated by adding `standalone=yes` to the gnu make command, for example ` make samtranslocations standalone=yes`.

### edit 'local.mk' (optional)

The a file **local.mk** can be created edited to override/add some definitions.

For example it can be used to set the HTTP proxy:

```
http.proxy.host=your.host.com
http.proxy.port=124567
```
## Source code 

[https://github.com/lindenb/jvarkit/tree/master/src/main/java/com/github/lindenb/jvarkit/tools/structvar/SamTranslocations.java](https://github.com/lindenb/jvarkit/tree/master/src/main/java/com/github/lindenb/jvarkit/tools/structvar/SamTranslocations.java)


<details>
<summary>Git History</summary>

```
Wed Nov 29 17:09:58 2017 +0100 ; adding samjdk / biostars answer ; https://github.com/lindenb/jvarkit/commit/9ed7b941944f653f5ca5cc822e069108ab8deaf6
Wed Nov 29 16:40:28 2017 +0100 ; sam transloc ; https://github.com/lindenb/jvarkit/commit/9b83677697adc04d1cd7411b2181fa933a47583d
```

</details>

## Contribute

- Issue Tracker: [http://github.com/lindenb/jvarkit/issues](http://github.com/lindenb/jvarkit/issues)
- Source Code: [http://github.com/lindenb/jvarkit](http://github.com/lindenb/jvarkit)

## License

The project is licensed under the MIT license.

## Citing

Should you cite **samtranslocations** ? [https://github.com/mr-c/shouldacite/blob/master/should-I-cite-this-software.md](https://github.com/mr-c/shouldacite/blob/master/should-I-cite-this-software.md)

The current reference is:

[http://dx.doi.org/10.6084/m9.figshare.1425030](http://dx.doi.org/10.6084/m9.figshare.1425030)

> Lindenbaum, Pierre (2015): JVarkit: java-based utilities for Bioinformatics. figshare.
> [http://dx.doi.org/10.6084/m9.figshare.1425030](http://dx.doi.org/10.6084/m9.figshare.1425030)


## converting to SVG

with the following XSLT stylesheet (last updated : 2017-11-30 )

```xslt
<?xml version='1.0' encoding="UTF-8"?>
<xsl:stylesheet xmlns:xsl='http://www.w3.org/1999/XSL/Transform' 
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xmlns:svg="http://www.w3.org/2000/svg"
	xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
        xmlns:str="http://exslt.org/strings"
        xmlns:math="http://exslt.org/math"
	extension-element-prefixes="str math"
	version='1.0'>
<xsl:output method="xml"  encoding="UTF-8"/>
<xsl:variable name="nsamples" select="count(translocations/partitions/partition)"/>
<xsl:variable name="cols" select="floor(math:sqrt($nsamples))"/>
<xsl:variable name="W" select="300"/>
<xsl:variable name="max-count">
  <xsl:for-each select="translocations/partitions/partition/event">
    <xsl:sort select="@count" data-type="number" order="descending" />
    <xsl:if test="position() = 1">
      <xsl:value-of select="number(@count)" />
    </xsl:if>
  </xsl:for-each>
</xsl:variable>



<xsl:template match="translocations">

<svg xmlns="http://www.w3.org/2000/svg">
  <xsl:attribute name="width">
  	<xsl:value-of select="$cols * $W"/>
  </xsl:attribute>
  <xsl:attribute name="height">
  	<xsl:value-of select="($cols +1) * $W"/>
  </xsl:attribute>
  <style>
        svg { fill:white;stroke:black;}
  	circle {fill:red; opacity:0.3; stroke:none;}
  	text {fill:gray;stroke:gray;font-size:10px;}
  	.c0 {fill:gray;opacity:0.05;}
  	.c1 {fill:white;opacity:0.05;}
  </style>
  <defs>
  	<g id="dict">
  		<xsl:apply-templates select="dictionary"/>
  	</g>
  </defs>
  <g>
  	<xsl:apply-templates select="partitions/partition"/>
  </g>
</svg>
</xsl:template>


<xsl:template match="dictionary">
<xsl:apply-templates select="contig"/>
<rect style="fill:none;stroke:gray;" x="0" y="0" width="{$W}"  height="{$W}">
</rect>
</xsl:template>

<xsl:template match="contig">
<xsl:variable name="clazz" select="concat('c',count(preceding-sibling::contig) mod 2)"/>
<xsl:variable name="genomelen" select="number(../@length)"/>
<xsl:variable name="x" select="(number(@index) div $genomelen) * $W"/>
<xsl:variable name="h" select="(number(@length) div $genomelen) * $W"/>

<rect class="{$clazz}"  x="0" y="{$x}" width="{$W}" height="{$h}">
</rect>
<rect class="{$clazz}"  y="0" x="{$x}" height="{$W}" width="{$h}">
</rect>
</xsl:template>


<xsl:template match="partition">
<xsl:variable name="idx" select="count(preceding-sibling::partition)"/>
<xsl:variable name="dx" select="($idx mod $cols) * $W"/>
<xsl:variable name="dy" select="floor(($idx div $cols)) * $W"/>
<g>
 <xsl:attribute name="transform">translate(<xsl:value-of select="$dx"/>,<xsl:value-of select="$dy"/>)</xsl:attribute>
 <title><xsl:value-of select="@name"/></title>
 <use x="0" y="0" href="#dict" />
 <text x="1" y="10"><xsl:value-of select="@name"/></text>
 <xsl:apply-templates select="event"/>
 
</g>
</xsl:template>

<xsl:template match="event">
<xsl:variable name="genomelen" select="number(../../../dictionary/@length)"/>
<xsl:variable name="cx" select="(number(start/@index) div $genomelen) * $W"/>
<xsl:variable name="cy" select="(number(end/@index) div $genomelen) * $W"/>
<xsl:variable name="r" select="0.1 + (number(@count) div $max-count) * 20.0"/>
<circle cx="{$cx}" cy="{$cy}" r="{$r}"/>
<circle cx="{$cy}" cy="{$cx}" r="{$r}"/>
</xsl:template>

</xsl:stylesheet>
```


